### ISSUE TITLE
Metabase renders the result but the query does not finish in the DB

### ISSUE THREAD

**Issue by paoliniluis** [metabase/metabase#39018]
### Describe the bug

Queries in the DB seem to continue in the background, while the query has already been finished for Metabase

### To Reproduce

1) add the Redshift dev DW to a Metabase (sample DB)
2) do the following query:
```
SELECT
  "public"."random_data"."text_column1" AS "text_column1",
  FLOOR(("public"."random_data"."numeric_column1" / 125.0)) * 125.0 AS "numeric_column1",
  FLOOR(("public"."random_data"."numeric_column4" / 125.0)) * 125.0 AS "numeric_column4",
  FLOOR(("public"."random_data"."numeric_column5" / 125.0)) * 125.0 AS "numeric_column5",
  FLOOR(("public"."random_data"."numeric_column8" / 125.0)) * 125.0 AS "numeric_column8",
  FLOOR(("public"."random_data"."numeric_column9" / 125.0)) * 125.0 AS "numeric_column9",
  FLOOR(
    ("public"."random_data"."numeric_column10" / 125.0)
  ) * 125.0 AS "numeric_column10",
  FLOOR(
    ("public"."random_data"."numeric_column13" / 125.0)
  ) * 125.0 AS "numeric_column13",
  COUNT(*) AS "count"
FROM
  "public"."random_data"
GROUP BY
  "public"."random_data"."text_column1",
  FLOOR(("public"."random_data"."numeric_column1" / 125.0)) * 125.0,
  FLOOR(("public"."random_data"."numeric_column4" / 125.0)) * 125.0,
  FLOOR(("public"."random_data"."numeric_column5" / 125.0)) * 125.0,
  FLOOR(("public"."random_data"."numeric_column8" / 125.0)) * 125.0,
  FLOOR(("public"."random_data"."numeric_column9" / 125.0)) * 125.0,
  FLOOR(
    ("public"."random_data"."numeric_column10" / 125.0)
  ) * 125.0,
  FLOOR(
    ("public"."random_data"."numeric_column13" / 125.0)
  ) * 125.0
ORDER BY
  "public"."random_data"."text_column1" ASC,
  FLOOR(("public"."random_data"."numeric_column1" / 125.0)) * 125.0 ASC,
  FLOOR(("public"."random_data"."numeric_column4" / 125.0)) * 125.0 ASC,
  FLOOR(("public"."random_data"."numeric_column5" / 125.0)) * 125.0 ASC,
  FLOOR(("public"."random_data"."numeric_column8" / 125.0)) * 125.0 ASC,
  FLOOR(("public"."random_data"."numeric_column9" / 125.0)) * 125.0 ASC,
  FLOOR(
    ("public"."random_data"."numeric_column10" / 125.0)
  ) * 125.0 ASC,
  FLOOR(
    ("public"."random_data"."numeric_column13" / 125.0)
  ) * 125.0 ASC
```
3) wait till it finishes and then...
4) go to Redshift and check the status of `select * from STV_INFLIGHT`, you'll see that the query is still running while in Metabase it has finished

### Expected behavior

The connection and query should finish if Metabase finishes

### Logs

NA

### Information about your Metabase installation

```JSON
v48.6
```

### Severity

P1

**Comment by lbrdnk** [comment]
It seems that we actually close the connection, but it takes much longer than returning the results back to the browser. I believe its a result of limiting rows to 2k.

After starting the query execution from GUI, we are able to check its status eg. using `select * from sys_query_history where end_time is null;`

That way we could acquire its id. We can further check its status using `select * from sys_query_history where query_id = <that id>;`

Query completes in the GUI in less than 30 seconds. Then there is a 30+ seconds window until the query's `end_time` column gets its value -- query completes on the Redshift's side.

Also executing this query through psql takes _those_ 60+ seconds.

---

_Now why is that?_ My hyphotesis: In limit postprocessing middleware we reduce only the `limit` number of rows ([limit.clj](https://github.com/metabase/metabase/blob/e9256fda39a3df843efacf4ce981f478317f7d6b/src/metabase/query_processor/middleware/limit.clj#L57)) -- in this case 2k. I believe that result set is batched, thus we are able to take those 2k rows and return those to the FE, while query is still running on the Redshift side.

Afterwards, during the qp machinery's return phase, result set is closed indeed, but the closing takes the delta between the time of seemingly completion in GUI and the actual completion on the Redshift side.

If I'm correct here, closing the result set (`with-open`'s finally) in `metabase.driver.sql-jdbc.execute/execute-reducible-query` ([execute.clj](https://github.com/metabase/metabase/blob/e9256fda39a3df843efacf4ce981f478317f7d6b/src/metabase/driver/sql_jdbc/execute.clj#L701)) blocks until the remaining results are computed, even though we reduce only first 2k.

To back that, measurements of `metabase.driver.sql-jdbc.execute/execute-reducible-query` follow:

```
"Elapsed time: 10037.283041 msecs"   -- timing of the statement execution
"Elapsed time: 513.398333 msecs"     -- timing of the respond call
"Elapsed time: 62449.817709 msecs"   -- timing of the call to with-open
```

Then, following is the log snippet, that's rendered _after the with-open returns_ -- long after FE has the results and considers the query completed.

```
2024-04-15 15:54:17,190 DEBUG middleware.log :: POST /api/dataset 202 [ASYNC: canceled] 1.0 mins (2 DB calls) App DB connections: 2/10 Jetty threads: 4/50 (2 idle, 0 queued) (101 total active threads) Queries in flight: 0 (0 queued); redshift DB 2 connections: 0/2 (0 threads blocked) {:metabase-user-id 1}
```

---

_What can we do about that?_ We could either kill the query instead of an attempt to close the result set or we could modify the limit middleware to actually modify the query so it contains the limit clause with the value of the actual limit, to avoid redundant computation on dbms side.

**Comment by perivamsi** [comment]
@paoliniluis do you have more info from the person who reported this? any ticket?

**Comment by mnherzz** [comment]
@perivamsi Apologies for the late response. It looks like this ended up getting fixed after one of the Metabase updates. We haven't had an issue for the past few weeks.

**Comment by perivamsi** [comment]
More context from the customer (copying this from an email conversation)

- We're running the latest version - 50.13. 
- 4 queries were left running. It seems to be an issue with window functions. We noticed that all 4 queries used the 'first_value' function with the window frame 'unbounded preceding and unbounded following'. I modified the query to use 'row_number', which initially resolved the issue. I added a few more join clauses to one of the modified queries and now that one is hanging again. 
- This query hangs when launched in the dashboard and when the question is run separately on its own outside of the dashboard. 
- The query times out after 10 minutes in Metabase. On the Redshift side, it never completes. The query stays open indefinitely until manually terminated.
- No, there aren't exceptions in Metabase logs
